build/build-report-components.js

```diff
@@ -0,0 +1,218 @@
+    elemToVarNames.set(el, varName);
+    return varName;
+  }
+
+  /**
+   * @param {Element} el
+   */
+  function process(el) {
+    const isSvg = el.namespaceURI && el.namespaceURI.endsWith('/svg');
+    const namespaceURI = isSvg ? el.namespaceURI : '';
+    const tagName = el.localName;
+    const className = el.classList.toString();
+
+    let createElementFnName = 'createElement';
+    const args = [tagName];
+    if (className) {
+      args.push(className);
+    }
+    if (namespaceURI) {
+      createElementFnName = 'createElementNS';
+      args.unshift(namespaceURI);
+    }
+
+    const varName = makeOrGetVarName(el);
+    lines.push(`const ${varName} = dom.${createElementFnName}(${serializeArguments(args)});`);
+
+    if (el.getAttributeNames) {
+      for (const attr of el.getAttributeNames() || []) {
+        if (attr === 'class') continue;
+
+        lines.push(`${varName}.setAttribute('${attr}', '${el.getAttribute(attr)}');`);
+      }
+    }
+
+    // const hasMultipleTextNodesWithContent = [...el.childNodes]
+    //   .map(n => n.nodeType === window.Node.TEXT_NODE &&
+    //             n.textContent && n.textContent.trim().length > 0).length > 1;
+    // const hasMultipleTextNodesWithContent = el.children.length > 2;
+    // for (const childEl of el.childNodes) {
+    //   const n = childEl;
+    //   // console.log(n.nodeType === window.Node.TEXT_NODE &&
+    //   //   n.textContent && n.textContent.trim());
+    //   if (n.nodeType === window.Node.TEXT_NODE && n.textContent && n.textContent.trim()) {
+    //     // console.log('s:', n.textContent);
+    //   }
+    // }
+    // console.log(el.children.length, hasMultipleTextNodesWithContent);
+
+    // Whitespace at the beginning and at the end can be removed.
+    let lower = 0;
+    let upper = el.childNodes.length;
+    if (el.childNodes.length > 0) {
+      const node = el.childNodes[0];
+      if (node.nodeType === window.Node.TEXT_NODE && node.textContent && !node.textContent.trim()) {
+        lower += 1;
+      }
+    }
+    if (el.childNodes.length > 1) {
+      const node = el.childNodes[el.childNodes.length - 1];
+      if (node.nodeType === window.Node.TEXT_NODE && node.textContent && !node.textContent.trim()) {
+        upper -= 1;
+      }
+    }
+
+    // TODO: the above assumes the first or last nodes wouldn't be a comment ... below is more generic. use?
+    // for (const node of el.childNodes) {
+    //   if (node.nodeType === window.Node.TEXT_NODE && node.textContent && !node.textContent.trim()) {
+    //     lower += 1;
+    //   } else {
+    //     break;
+    //   }
+    // }
+    // for (const node of [...el.childNodes].reverse()) {
+    //   if (node.nodeType === window.Node.TEXT_NODE && node.textContent && !node.textContent.trim()) {
+    //     upper -= 1;
+    //   } else {
+    //     break;
+    //   }
+    // }
+
+    for (const childNode of [...el.childNodes].slice(lower, upper)) {
```

Comment:
I noticed that this doesn't catch small differences in whitespace.... ex, for `footer` the "Generated by Lighthouse" looks like this for both generated and "original", according to jsdom:

```
<div class="lh-env"><div class="lh-env__title">Runtime Settings</div><ul class="lh-env__items"></ul></div><div class="lh-generated">
          Generated by <b>Lighthouse</b><span class="lh-footer__version"></span> |
          <a class="lh-footer__version_issue" href="https://github.com/GoogleChrome/Lighthouse/issues" target="_blank" rel="noopener">File an issue</a></div>
```

Note the lack of a space after `Lighthouse</b>`. Luckily, a test in report-renderer caught this  error with a lucky regex.

As for fixing it... The reason it's missed is because text nodes of just whitespace are skipped. 99% of them are pointless. So if we want to keep it, we need a way to know when the text is not pointless. The first and last nodes are typically whitespace, so I am explicitly checking for those (and assuming all-ws-nodes in the middle are important). Then, I allowed all other text nodes regardless of contents. This adds a lot of unnecessary text nodes because of ...

```
<li class="lh-env__item">
    <span class="lh-env__name"></span>             (LOL i am become text node " ")
    <span class="lh-env__description"></span>
  </li>


/**
 * @param {DOM} dom
 */
function createEnvItemComponent(dom) {
  const v0 = dom.document().createDocumentFragment();
  const v1 = dom.createElement('li', 'lh-env__item');
  const v2 = dom.createElement('span', 'lh-env__name');
  v1.append(v2);
  v1.append(dom.document().createTextNode(' ')); // hi
  const v3 = dom.createElement('span', 'lh-env__description');
  v1.append(v3);
  v0.append(v1);
  return v0;
}
```

What I've just pushed is this process.

Instead, maybe `Generated by <b>Lighthouse</b> <span class="lh-footer__version"></span>` could be tweaked to something else... but we might slip up somewhere else on the same issue.

Maybe better: check if a text node has a `span` plus some other element for a sibling, if it does allow all-whitespace, otherwise don't.

Reply:
Alright, this was much simpler with the sibling check. Here's the diff between the original pass (ignore every all-ws-text-node) and the sibling check:

```
diff --git a/report/renderer/components.js b/report/renderer/components.js
index a638a4508..fc840bff4 100644
--- a/report/renderer/components.js
+++ b/report/renderer/components.js
@@ -21,13 +21,14 @@ function create3pFilterComponent(dom) {
   v4.setAttribute('type', 'checkbox');
   v4.setAttribute('checked', '');
   v3.append(v4);
+  v3.append(dom.document().createTextNode(' '));
   const v5 = dom.createElement('span', 'lh-3p-ui-string');
   v5.append(dom.document().createTextNode('Show 3rd party resources'));
   v3.append(v5);
   v3.append(dom.document().createTextNode(' ('));
   const v6 = dom.createElement('span', 'lh-3p-filter-count');
   v3.append(v6);
-  v3.append(dom.document().createTextNode(')\n    '));
+  v3.append(dom.document().createTextNode(') '));
   v2.append(v3);
   v0.append(v2);
   return v0;
@@ -44,12 +45,15 @@ function createAuditComponent(dom) {
   const v4 = dom.createElement('div', 'lh-audit__header lh-expandable-details__summary');
   const v5 = dom.createElement('span', 'lh-audit__score-icon');
   v4.append(v5);
+  v4.append(dom.document().createTextNode(' '));
   const v6 = dom.createElement('span', 'lh-audit__title-and-text');
   const v7 = dom.createElement('span', 'lh-audit__title');
   v6.append(v7);
+  v6.append(dom.document().createTextNode(' '));
   const v8 = dom.createElement('span', 'lh-audit__display-text');
   v6.append(v8);
   v4.append(v6);
+  v4.append(dom.document().createTextNode(' '));
   const v9 = dom.createElement('div', 'lh-chevron-container');
   v4.append(v9);
   v3.append(v4);
@@ -109,6 +113,7 @@ function createClumpComponent(dom) {
   const v4 = dom.createElement('div', 'lh-audit-group__header');
   const v5 = dom.createElement('span', 'lh-audit-group__title');
   v4.append(v5);
+  v4.append(dom.document().createTextNode(' '));
   const v6 = dom.createElement('span', 'lh-audit-group__itemcount');
   v4.append(v6);
   v3.append(v4);
@@ -131,6 +136,7 @@ function createCrcComponent(dom) {
   const v4 = dom.createElement('div', 'lh-crc__summary-value');
   const v5 = dom.createElement('span', 'lh-crc__longest_duration_label');
   v4.append(v5);
+  v4.append(dom.document().createTextNode(' '));
   const v6 = dom.createElement('b', 'lh-crc__longest_duration');
   v4.append(v6);
   v3.append(v4);
@@ -151,6 +157,7 @@ function createCrcChainComponent(dom) {
   const v1 = dom.createElement('div', 'crc-node');
   const v2 = dom.createElement('span', 'crc-node__tree-marker');
   v1.append(v2);
+  v1.append(dom.document().createTextNode(' '));
   const v3 = dom.createElement('span', 'crc-node__tree-value');
   v1.append(v3);
   v0.append(v1);
@@ -192,6 +199,7 @@ function createEnvItemComponent(dom) {
   const v1 = dom.createElement('li', 'lh-env__item');
   const v2 = dom.createElement('span', 'lh-env__name');
   v1.append(v2);
+  v1.append(dom.document().createTextNode(' '));
   const v3 = dom.createElement('span', 'lh-env__description');
   v1.append(v3);
   v0.append(v1);
@@ -215,13 +223,14 @@ function createFooterComponent(dom) {
   v3.append(v5);
   v2.append(v3);
   const v6 = dom.createElement('div', 'lh-generated');
-  v6.append(dom.document().createTextNode('\n      Generated by '));
+  v6.append(dom.document().createTextNode(' Generated by '));
   const v7 = dom.createElement('b');
   v7.append(dom.document().createTextNode('Lighthouse'));
   v6.append(v7);
+  v6.append(dom.document().createTextNode(' '));
   const v8 = dom.createElement('span', 'lh-footer__version');
   v6.append(v8);
-  v6.append(dom.document().createTextNode(' |\n      '));
+  v6.append(dom.document().createTextNode(' | '));
   const v9 = dom.createElement('a', 'lh-footer__version_issue');
   v9.setAttribute('href', 'https://github.com/GoogleChrome/Lighthouse/issues');
   v9.setAttribute('target', '_blank');
@@ -405,6 +414,7 @@ function createMetricComponent(dom) {
   const v2 = dom.createElement('div', 'lh-metric__innerwrap');
   const v3 = dom.createElement('span', 'lh-metric__title');
   v2.append(v3);
+  v2.append(dom.document().createTextNode(' '));
   const v4 = dom.createElement('div', 'lh-metric__value');
   v2.append(v4);
   const v5 = dom.createElement('div', 'lh-metric__description');
@@ -467,6 +477,7 @@ function createOpportunityComponent(dom) {
   const v6 = dom.createElement('div', 'lh-load-opportunity__col lh-load-opportunity__col--one');
   const v7 = dom.createElement('span', 'lh-audit__score-icon');
   v6.append(v7);
+  v6.append(dom.document().createTextNode(' '));
   const v8 = dom.createElement('div', 'lh-audit__title');
   v6.append(v8);
   v5.append(v6);
@@ -517,9 +528,11 @@ function createScorescaleComponent(dom) {
   const v2 = dom.createElement('span', 'lh-scorescale-range lh-scorescale-range--fail');
   v2.append(dom.document().createTextNode('0–49'));
   v1.append(v2);
+  v1.append(dom.document().createTextNode(' '));
   const v3 = dom.createElement('span', 'lh-scorescale-range lh-scorescale-range--average');
   v3.append(dom.document().createTextNode('50–89'));
   v1.append(v3);
+  v1.append(dom.document().createTextNode(' '));
   const v4 = dom.createElement('span', 'lh-scorescale-range lh-scorescale-range--pass');
   v4.append(dom.document().createTextNode('90–100'));
   v1.append(v4);
@@ -586,6 +599,7 @@ function createSnippetHeaderComponent(dom) {
   const v4 = dom.createElement('button', 'lh-snippet__toggle-expand');
   const v5 = dom.createElement('span', 'lh-snippet__btn-label-collapse lh-snippet__show-if-expanded');
   v4.append(v5);
+  v4.append(dom.document().createTextNode(' '));
   const v6 = dom.createElement('span', 'lh-snippet__btn-label-expand lh-snippet__show-if-collapsed');
   v4.append(v6);
   v1.append(v4);

```